### ISSUE TITLE
Don't force `keyof` constraint for `Ember.get`

### ISSUE THREAD

**Issue by peabnuts123** [typed-ember/ember-cli-typescript#791]
### Which package(s) does this enhancement pertain to?
  - [x] @types/ember__object

### Please write a user story for this feature

I think that the current type definitions for `get` break compatibility with Ember, for the ability to pull off what seems to be just a bit of a neat trick. This applies to both forms `this.get` and `get` imported from `@ember/object`. 

The current type definition is as follows:

```typescript
export function get<T, K extends keyof T>(
    obj: T,
    key: K
): UnwrapComputedPropertyGetter<T[K]>;
```

This utilises TypeScript's `keyof` keyword to enforce that whatever string you pass into `get()` is a compile-time known property name of `obj` (or `this` if using `this.get()`). In certain use-cases, this gives better type-safety with your code. Particularly, this favours older projects on older versions of Ember, using Ember-objects instead of ES6 classes, etc. 

However, the more I use this, the more it feels like it doesn't match Ember, and the more it is becoming something I need to avoid in order for my code to work (rather than something I should embrace), and I think it's worth considering changing the definitions "back". I understand significant discussion occurred when this library was less mature than it is now. 

From my perspective, the semantics as they are today do not match Ember's JavaScript semantics in the following ways:
  - Non-existent keys are a compiler error, instead of returning `undefined`
  - Paths are "non-existent" keys - they cannot be used at all
  - Cannot dereference an object with a variable unless its type is `keyof T` e.g. `this.get(propName)`
  - Essentially there is no way to determine if an object has a property named `propName` using `get()`
    - This is a pretty common pattern in traditional JavaScript code e.g. checking if a hook is defined by name, or something
    - This breaks compatibility with the design of TypeScript, see also: TypeScript method overloading

These differences seem pretty fundamental, the type definition for `get` does not very closely match the semantics of the implementation at all. The only use-case left is to use it to request a known property by name that you know exists at compile time. This seems like a trivial use-case and it seems a bit overboard to sacrifice all other use-cases for the sake of a strongly typed result in this particular use-case (granted, this is probably the case 80% of the time). I've had to convert many uses of `get()` to lodash's `_.get()` - I simply have no other way of doing it?

From my perspective, a simpler type definition would be a lot more useful: 

```TypeScript
function get<T>(
    obj: T,
    key: string
): any | undefined;
```

This is a "worst case" example i.e. the most simple example, for demonstrative purposes. There are obviously issues with returning `any`, I think you could write a simple typedef that also mitigated these issues to some degree. Ultimately, I think fully embracing TypeScript in Ember should take the form of using a version of Ember that is new-enough to support ES5 getters/setters such that you can / should use `this.myThing` instead of `this.get('myThing')`. This way the 80% use-case is still strongly-typed and working, but there is room for less-common use-cases, and for code that came from JavaScript to be ported to TypeScript (at a cost of still being kind of wonky - you are encouraged to refactor it to fully utilise TypeScript once you have the ability to).

To summarise, it seems that right now the typedefs for `get()` are over-optimised for the most basic use-case, at the cost of less-common use-cases being essentially unusable. Particularly, some JavaScript ported to TypeScript faces great difficulty, which should be a priority for this library to support. The type semantics are heavily mismatched from the implementation semantics in a way that some JavaScript code cannot be ported to TypeScript. I think a simpler typedef should be used.

**Comment by chriskrycho** [comment]
Thanks for the thoughtful writeup! I'm working on a careful response; you can expect to see that by the end of the day today. 😄

**Comment by chriskrycho** [comment]
I've broken this response into three parts:

- a workaround for the issue you’re encountering
- a detailed explanation for why we chose the types we’ve chosen
- some specific point-by-point responses to a few of your comments

## Workaround

In general, you *should* be able to solve the immediate problem by treating the type you’re working with as a dictionary:

```ts
interface Dict<T> {
  [key: string]: T | undefined
}
```

You can use this code sample with the above definition to see that you can safely treat any `EmberObject` as a dictionary:

```ts
import EmberObject, { get } from '@ember/object';

function treatItLikeADict(someObj: Dict<unknown>): unknown | undefined {
  return get(someObj, 'foo');
}

let anyEmberObject = EmberObject.create({
  foo: 'hey',
  bar: 'bye',
});

treatItLikeADict(anyEmberObject);
```

Similarly, in the context of any `EmberObject`-descended class, you can safely cast `this as Dict<T>`:

```ts
import Service from '@ember/service';
import { get } from '@ember/object';

export default class MyService extends Service {
  someMethod(): unknown {
    return get(this as Dict<unknown>, 'foo');
  }
}
```

That is of course slightly more boilerplate than having `get` *always* return `any`, but as I’ll describe below, this is a *necessary* tradeoff. If you decide you *actually* want this to be the default behavior everywhere, you can easily opt into it by adding this to your app at `/types/ember__object/index.d.ts`:

```ts
import '@ember/object'; // makes declaration merging work

declare module '@ember/object' {
  export function get(obj: any, key: string): any; // or `unknown`
}
```

However, once you do this, you will get *no* type-checking around your use of `get`. That may be fine in your codebase, or it may not. It’s strictly identical semantically with everything Ember’s `get` helper *allows*… but it’s not normally what people actually *want*. In any case, read on for an explanation of why it *isn’t* fine for the default!

## Explanation

First, and to be quite direct: what we’ve chosen does *not* "break compatibility with Ember". I’ll explain why in detail below, but I think it’s important to recognize that "doesn’t work the way I want it to" is not the same as "broken"—and there are *very* good reasons for the defaults we’ve chosen here.

To begin with, what we’re doing with `get` isn’t just a *neat trick*, as you put it. Instead, it represents a decision about how to handle a fundamental (and inescapable!) tension between two different ways people tend to approach this space:

- treating JS objects as arbitrary bags of data, i.e. effectively as *dictionaries* or *maps* which can have any key at any time—where accessing an unknown property is *expected*

- treating JS objects as carefully-structured *types*, which have known properties—where accessing an *unknown* property is usually a programming error

It’s not possible for one set of type definitions to cover *both* of the cases for the *same* object, though it *is* possible for types to correctly handle both if you’re able to distinguish which type you’re dealing with. I also think both are perfectly valid strategies to use, and by and large we try very hard to support both, with a bias toward supporting the one that is *harder* to opt into as a user of the type definitions.

That is: our types need to support either the carefully-structured model with user-driven *loosening* into the bag-of-data model, or the bag-of-data case with user-driven *tightening* into the carefully-structured model. And it’s much, *much* harder to get the types correct for the carefully-structured model, and quite straightforward in that case to provide ways for users to *loosen* when appropriate.

When we started working on these types in 2017, exactly what you have proposed is what existed, and *the experience was terrible*. The reason is that at that time, you *had* to use `get` for everything, and that meant that you could never have *any* info from the types about what kind of property you were accessing.

This meant that for the 98% case, when writing code like this, you got *zero help*:

```ts
import Service from '@ember/service';
import { get, computed } from '@ember/object';

export default class extends Service {
  foo = 'hello';
  
  @computed('foo')
  get bar() {
    // whoops, typo'd don't find out till runtime!
    // why do I even *have* a type system?!?
    return this.get('foi');
  }
});
```

That’s a simple example, but it’s one that not only mattered when we started, but *still matters* to a lot of people today—if this doesn’t work, then you can’t use TypeScript in a type-safe way anywhere that Proxies are in use, full stop. While it may be surprising to you, "anywhere that Proxies are in use" covers a *lot* of places in many codebases, not least because of long-standing async defaults around the behavior of `hasMany` and `belongsTo` in Ember Data.

What’s more, unlike the easy loosening I demonstrated in the workaround section above, there’s no way to opt *back into* this type-checking if we supplied the default `get` you suggested. Having the loosest-possible variant *anywhere* would mean that TypeScript always resolved to it *everywhere*.

The net of all of this is that the choice we’ve made is the *only* choice which actually supports both real-world use cases, *allowing* people to opt out of type safety in these scenarios if they wish while *enabling* type safety for those who do in fact want it. Unsurprisingly, in our experience, the majority of people converting their codebases to TypeScript *do* want that type safety around these kinds of lookups: that’s why they’re converting!

## Responses

Finally, a few specific comments on statements you made.

> - Non-existent keys are a compiler error, instead of returning `undefined`

I believe this is covered in detail above.

> - Paths are "non-existent" keys - they cannot be used at all

This is indeed unfortunate, and it’s one reason we’re *very* eager to help people move into the Octane programming model. (The nested `get` is literally my *very* least favorite thing about a legacy codebase converted to TS.) However, it’s also a tradeoff, for all the reasons outlined above, because this is a valid JS object:

```js
const foo = {
  'bar.baz.quux': 'Hi'
}
```

While no one *should* write that, people *do* write that, and supporting type-safe lookups *at all* means not handling that as a path. Happily, as people *do* migrate into TS and especially Octane, it becomes possible to eliminate these. In fact, in most cases, people currently adopting TS can *also* switch away from using `get` in this circumstance at all—per your point about modern Ember semantics.

> - Cannot dereference an object with a variable unless its type is `keyof T` e.g. `this.get(propName)`

Yep, and for certain dynamic programming patterns this is indeed quite painful! However, the same workaround suggested above also works here.

> - Essentially there is no way to determine if an object has a property named `propName` using `get()`
>     - This is a pretty common pattern in traditional JavaScript code e.g. checking if a hook is defined by name, or something
>     - This breaks compatibility with the design of TypeScript, see also: TypeScript method overloading

This is partially correct in the first case, incorrect in the second. (Method overloading works *just fine* with our types, and we use it heavily in our types and in codebases I’ve worked on.) In the first case, though, there are two other important points to consider:

1. There are straightforward workarounds using [type narrowing](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types), when the dynamic check-and-operate-if-present approach is actually what you need.

2. Just as importantly, migrating into TS often allows you to *stop* doing a lot of that at runtime, because you can check it at compile time instead. This means that some of the patterns you’re used to are simultaneously less elegant *and* less necessary in TS.

> This seems like a trivial use-case and it seems a bit overboard to sacrifice all other use-cases for the sake of a strongly typed result in this particular use-case (granted, this is probably the case 80% of the time).



> There are obviously issues with returning `any`, I think you could write a simple typedef that also mitigated these issues to some degree.

You seem to be suggesting that it’s worth letting the 20% case trump the 80% case, *and* that it’s trivial to invert how the types work. Let’s just say—from everyone on the team’s experience!—that if you spend the time to get a "simple" typedef working to a degree that’s actually satisfactory for any large existing codebase, you’ll find that it is in fact *not at all simple*. (It ends up looking like, well… the existing type definition.)

What’s more, emphasizing the 20% case (which *does* have good workarounds) at the *expense* of the 80% case (which does *not*) is a design choice that simply doesn’t make sense to us.

But of course all of that is moot: as I described above, what you’re suggesting is actually impossible in a single set of type definitions.

> Ultimately, I think fully embracing TypeScript in Ember should take the form of using a version of Ember that is new-enough to support ES5 getters/setters such that you can / should use `this.myThing` instead of `this.get('myThing')`.

It's fine if that’s how you want to work. And we’re actively working on ways to make it so that users can opt into lighter-weight and easier versions of the type definitions if they’re 100% in the Octane programming model. However, while *you* may be comfortable with that model, our job is to support *all* of the Ember community, not just folks who can be in that position. There are codebases out there which are *years* away from being at that point, and they *cannot* adopt TypeScript unless we provide support for migrating gradually.

> The type semantics are heavily mismatched from the implementation semantics in a way that some JavaScript code cannot be ported to TypeScript.

As noted above: this is not the case. The type semantics *exactly* match the more-difficult-to-capture scenario, and can be loosened as shown in the example code above.

> ```ts
> function get<T>(
>     obj: T,
>     key: string
> ): any | undefined;
> ```

Note that `any | undefined` is identical with `any`; this type definition is not actually saying what you think it is. Additionally, making `T` generic here doesn’t actually accomplish anything. The correct type definition (as given above) is:

```ts
function get(obj: any, key: string): any;
```

Notice that this does in fact give you *exactly* the semantics you specified wanting: it’s precisely as loose as Ember’s `get` will allow. If you’d like to call it as `get(42, 'waffles')`, you will indeed get back `undefined`!

---

Note that I'm closing this issue, since we won't be making this change, but feel free to continue the conversation!

**Comment by peabnuts123** [comment]
Thanks for the detailed response, Chris! I did not mean to take such a chunk of your time. I think that you've interpreted some of my points as more harsh or more specific than they actually are, such as:

  - Suggesting the 20% case should trump the 80% case
  - Implying there was not already a lot of discussion on this topic 😅
  - Suggesting that `get()` _should_ return `any`
  - Suggesting that people on older projects should just upgrade
  - Suggesting that it's easy to write a type definition that covers both scenarios you described above
  - Suggesting I can't easily work-around the issue with my own type definitions / utility functions / alternative functions
  - Generally asserting that my own programming patterns should be how everybody writes their projects / are the only use-cases I had considered

I'm happy for this issue to be closed, I just wanted to raise it and at the least give this discussion a (another?) home. 

To continue the conversation for the sake of clarity: I don't think any of the points listed above, apologies if the subtleties of what I was saying did not come through. I appreciate the notion of loose types (with the ability to tighten your types - essentially putting the burden on the consumer) vs. tight types (with the ability to loosen - essentially putting the burden on ember-cli-typescript), and I agree that the chosen paradigm (tight, with optional loosening) is the right approach to favour.

I think the example you've given is a good one:

```TypeScript
import Service from '@ember/service';
import { get, computed } from '@ember/object';

export default class MyService extends Service {
  foo = 'hello';
  
  @computed('foo')
  get bar() {
    // whoops, typo'd don't find out till runtime!
    // why do I even *have* a type system?!?
    return this.get('foi');
  }
}
```

This point is a demonstration of one design decision that I personally disagree with or perhaps otherwise lack some other information about. If this example were re-written to use `this.foi` this type-error would be caught. Native ES5 getters were generally available in Ember 3.1 which was released over a year ago. This is not to say anything along the lines of "people should upgrade to Ember 3.1 and then just use that" – it's merely an observation to say that there was a decision made that projects on versions of Ember less than 3.1 are more important or more generally supported than newer versions. It's important to support as many versions and codebases as is reasonable, but it seems to me that more modern projects should be the most supported. Admittedly I am not wholly across the changes happening in Octane so I don't know what you're referring to when you say `[lack of support for paths is] one reason we’re very eager to help people move into the Octane programming model`.

I can clearly work around these issues in my own way. For example, I suggested that I have been using Lodash's `_.get()` instead which has exactly the super-generic type-def you've listed above: 
```TypeScript
get(
    object: any,
    path: PropertyPath,
    defaultValue?: any
): any;
```
I created this issue because it felt like this design pattern was encouraging me to work around its own design - that my old JavaScript code using Ember had to be refactored to use Lodash in order to compile now. It will take a long time for my current main project (4+ years of Ember JS code)  to be refactored enough to hit a critical mass of TypeScript.

I don't want to write another huge wall of text so I'll leave it there for now, I hope this has made more sense. Ultimately I would love for this issue to reach a point of either inciting change in this library (I know this is pretty unlikely) or myself understanding exactly what opinions and decisions are held by the maintainers here that I don't personally agree with - which is totally fine! I'm just one person with my own ideas and viewpoints. Again thanks for the carefully thought-out and detailed response! I appreciate the work you (Chris, and the rest of the ember-typescript community) have done.

**Comment by chriskrycho** [comment]
No worries; I was not offended! Just wanted to be very clear about the reasoning and tradeoffs in play.

> If this example were re-written to use `this.foi` this type-error would be caught. Native ES5 getters were generally available in Ember 3.1 which was released over a year ago. This is not to say anything along the lines of "people should upgrade to Ember 3.1 and then just use that" – it's merely an observation to say that there was a decision made that projects on versions of Ember less than 3.1 are more important or more generally supported than newer versions.

The trick is that while indeed 3.1 made direct access available in many cases, the exceptions *still* require support. Many things which touch Ember Data *do* have to use `this.get` even today, for example, and would lose type-safety without what we support here. Likewise, any class which uses `Ember.Proxy` would have the same set of problems for proxied properties.

To be clear: if we could safely drop all support for `.get` today and/or just type it without regard for safety because it was universally unnecessary for people to do type-safe `this.get` even in a fully modern Ember application, we'd be delighted. The problem is that we *can't* until the situation changes around some of those `Proxy`-reliant implementations.

As a last point: we're *exceedingly* unlikely to do anything that introduces `any` into people's builds without a breaking change in the type definitions, for the simple reason that doing so may be "safe" in the sense that people's code will keep building but it's *incredibly* unsafe and indeed quite hostile to users who may have done a great deal of work to *eliminate* `any` from their code.

For all those reasons, our take is that we will continue supporting type-safe lookups with `get` as long as core, supported parts of the Ember development experience (`Proxy` and especially Ember Data's use thereof) depend on it, and that the workarounds available mean that people like you who *want* to opt into an unsafe-by-default pattern can freely do so. That enables all Ember users to adopt TS as makes sense for them and their apps and addons.

One final note: that module merging example I shared works exactly the way you want! Given the way you're working on this conversion and the shape of the codebase you've alluded to, I think you probably *should* use that to eliminate an adoption pain point, assuming you're able to use `this.foo` lookups otherwise. That'll mean you don't need to rewrite to use `_.get` just to get type-checking with the behavior you expect, and it won't be any worse type-safety-wise than doing so!
